modules/ding_gatewayf/modules/ding_gatewayf_registration/ding_gatewayf_registration.module

```diff
@@ -0,0 +1,456 @@
+  );
+
+  // Define after create info page.
+  $form['ding_gatewayf_registration']['success'] = array(
+    '#type' => 'fieldset',
+    '#title' => t('Successful registration page'),
+    '#description' => t('Information displayed after the user have been successful registered at the provider.'),
+    '#tree' => TRUE,
+    '#collapsible' => TRUE,
+    '#collapsed' => TRUE,
+  );
+
+  $form['ding_gatewayf_registration']['success']['title'] = array(
+    '#type' => 'textfield',
+    '#title' => t('Title'),
+    '#default_value' => $defaults['success']['title'],
+  );
+
+  $form['ding_gatewayf_registration']['success']['content'] = array(
+    '#type' => 'textarea',
+    '#title' => t('Content'),
+    '#description' => t('This field accept HTML.'),
+    '#default_value' => $defaults['success']['content'],
+  );
+}
+
+/**
+ * Generate render array for the information page.
+ *
+ * @return array
+ *   Render array with the page content.
+ */
+function ding_gatewayf_registration_information() {
+  $config = _ding_gatewayf_registration_load_config();
+
+  // Set page title to the same as the pages headline.
+  drupal_set_title($config['information']['title']);
+
+  return array(
+    '#theme' => 'ding_gatewayf_registration_information',
+    '#title' => $config['information']['title'],
+    '#content' => $config['information']['content'],
+    '#link' => array(
+      '#theme' => 'link',
+      '#text' => $config['information']['link'],
+      '#path' => DING_GATEWAYF_LOGIN_URL,
+      '#options' => array(
+        'attributes' => array(
+          // This class "form-submit" is used to make the link look and behave
+          // link a button.
+          'class' => array(
+            'form-submit',
+            'ding-gatewayf-link',
+          ),
+        ),
+        'query' => array(
+          'destination' => DING_GATEWAYF_REGISTRATION_ACCEPTANCE_URL,
+        ),
+        'html' => TRUE,
+      ),
+    ),
+  );
+}
+
+/**
+ * Implements hook_form().
+ */
+function ding_gatewayf_registration_acceptance_form() {
+  $config = _ding_gatewayf_registration_load_config();
+
+  // This is a little style hack to get around the panels design requirements of
+  // the DDB default theme. The module existing_pages could have been used to
+  // wrapper the form in panels. But seen form a performance point this is
+  // faster and simpler.
+  $form = array(
+    '#prefix' => '<div class="empty-sidebars default-layout"><div class="layout-wrapper"><div class="pane-content"><div class="primary-content">',
+    '#suffix' => '</div></div></div></div>',
+  );
+
+  $form['accept'] = array(
+    '#type' => 'checkbox',
+    '#title' => t('I accept the user agreements'),
+    '#description' => $config['acceptance']['description'],
+    '#required' => TRUE,
+  );
+
+  $form['name'] = array(
+    '#type' => 'textfield',
+    '#title' => t('Name'),
+    '#description' => t('You full name.'),
+    '#maxlength' => 100,
```

Comment:
@@ -0,0 +1,402 @@
 +<?php
 +/**
 + * @file
 + * Handles user creation in the provider with information from WAYF login.
 + */
 +
 +/**
 + * Implements hook_permission().
 + */
 +function ding_wayf_dk_create_user_permission() {
 +  return array(
 +    'administer wayf user creation' => array(
 +      'title' => t('Administer WAYF user creation'),
 +    ),
 +  );
 +}
 +
 +/**
 + * Implements hook_menu().
 + */
 +function ding_wayf_dk_create_user_menu() {
 +  $items = array();
 +
 +  $items['wayf/create-user/before-page'] = array(
 +    'title' => 'Create new user information page',
 +    'description' => 'Create new user using WAYF login.',
 +    'page callback' => 'ding_wayf_dk_create_user_info_page',
 +    'access callback' => TRUE,
 +    'type' => MENU_CALLBACK,
 +  );
 +
 +  $items['wayf/create-user/notice'] = array(
 +    'title' => 'User creation notice',
 +    'description' => 'Notice about account creation.',
 +    'page callback' => 'ding_wayf_dk_create_user_notice',
 +    'access callback' => TRUE,
 +    'type' => MENU_CALLBACK,
 +  );
 +
 +  $items['wayf/create-user/form'] = array(
 +    'title' => 'User creation form',
 +    'description' => 'Form used to account creation.',
 +    'page callback' => 'drupal_get_form',
 +    'page arguments' => array('ding_wayf_dk_create_user_form'),
 +    'access callback' => TRUE,
 +    'type' => MENU_CALLBACK,
 +  );
 +
 +  return $items;
 +}
 +
 +/**
 + * Implements hook_ctools_plugin_directory().
 + */
 +function ding_wayf_dk_create_user_ctools_plugin_directory($owner, $plugin_type) {
 +  return 'plugins/' . $plugin_type;
 +}
 +
 +/**
 + * Implements hook_wayf_dk_login_authenticate_result_alter().
 + */
 +function ding_wayf_dk_create_user_wayf_dk_login_authenticate_result_alter($result) {
 +  // After the user have logged in using WAYF this alter hook is called. If the
 +  // destination url (redirect) in the session (created at login) is the create
 +  // new user path. This path is part of the "create user" link and then we know
 +  // that this is a create use request.
 +  if (isset($_SESSION['ding_wayf_dk_login_destination']) && $_SESSION['ding_wayf_dk_login_destination'] == 'wayf/create-user/form') {
 +    // Extract unique id form the WAYF attributes (the CPR number).
 +    preg_match('/\d{10}$/', reset($result['attributes']['schacPersonalUniqueID']), $unique_id);
 +    $unique_id = reset($unique_id);
 +
 +    // Check if age limit should be set.
 +    $limit = variable_get('ding_wayf_user_create_age_limit', '');
 +    if (!empty($limit) && is_numeric($limit)) {
@kasperg   Owner
kasperg added a note 10 days ago
Validation in the settings form should ensure that $limit is numeric.
 Add a line note
 +
 +      // Magic function to get the 1900, 1800 or 2000 prefixes for year.
@kasperg   Owner
kasperg added a note 10 days ago
1800-prefix. Really?
 @cableman   Owner
cableman added a note 4 days ago
Yes, it's an standard function to detect CPR numbers, so for completeness.
 Add a line note
 +      // @see http://da.wikipedia.org/wiki/CPR-nummer
 +      $year = substr($unique_id, 4, 2);
 +      switch (substr($unique_id, 6, 1)) {
 +        case 0:
 +        case 1:
 +        case 2:
 +        case 3:
 +          $year = '19' . $year;
 +          break;
 +
 +        case 4:
 +        case 9:
 +          $year = $year > 36 ? '19' . $year : '20' . $year;
 +          break;
 +
 +        case 5:
 +        case 6:
 +        case 7:
 +        case 8:
 +          $year = $year > 57 ? '18' . $year : '20' . $year;
 +          break;
 +      }
 +
 +      // Create date time object to use it's diff method.
 +      $from = new DateTime();
 +      $from->setDate($year, substr($unique_id, 2, 2), substr($unique_id, 0, 2));
 +      $to = new DateTime('tomorrow');
@kasperg   Owner
kasperg added a note 10 days ago
Why tomorrow instead of now?
 @cableman   Owner
cableman added a note 4 days ago
It's due to the time of day, when we created this months ago we did a lot of testing and this was the correct why to do it.

We don't have the test code any more and can't remember the why, but it's something about the day they turn (the limit) they should be able to be created.
 Add a line note
 +
 +      // Check if limit have been reached.
 +      if ($from->diff($to)->y < $limit) {
 +        // The age limit has to kick in.
 +        drupal_set_message(t('You are not allowed to automatically be created because of the age limit (!limit years). Please go to your local library to registry.', array('!limit' => $limit)));
 +
 +        // Redirect to front page.
 +        drupal_goto('<front>');
 +      }
 +    }
 +
 +    // Store the schacPersonalUniqueID (CPR) in the current session to use when
 +    // sending the "user create" request to the provider in the redirect to
 +    // create new user form. It's only temporally store in the session and
 +    // removed as soon as possible.
 +    $_SESSION['ding_wayf_dk_create_user'] = $unique_id;
 +  }
 +}
 +
 +/**
 + * Implements hook_form_wayf_dk_login__settings_form_alter().
 + */
 +function ding_wayf_dk_create_user_form_wayf_dk_login__settings_form_alter(&$form, &$form_state, $form_id) {
 +  $form['user'] = array(
 +    '#type' => 'fieldset',
 +    '#title' => t('Create user'),
 +    '#group' => 'tabs',
 +  );
 +
 +  $form['user']['ding_wayf_user_create_age_limit'] = array(
 +    '#type' => 'textfield',
 +    '#title' => t('Age limit'),
 +    '#description' => t('The age limit will be used to ensure that no user is created without begin above the limit. If left empty no age limit is enforced.'),
 +    '#default_value' => variable_get('ding_wayf_user_create_age_limit', ''),
 +  );
 +
 +  // Define before create info page.
 +  $form['user']['ding_wayf_user_create_before_info'] = array(
 +    '#type' => 'fieldset',
 +    '#tree' => TRUE,
 +    '#collapsible' => TRUE,
 +    '#collapsed' => TRUE,
 +    '#title' => t('Information page (before)'),
 +    '#description' => t('Information to display before the user clicks on the "create user" link before the user is asked to login with WAYF.'),
 +  );
 +
 +  $defaults = variable_get('ding_wayf_user_create_before_info', array());
 +  $form['user']['ding_wayf_user_create_before_info']['title'] = array(
 +    '#type' => 'textfield',
 +    '#title' => t('Title'),
 +    '#default_value' => isset($defaults['title']) ? $defaults['title'] : '',
 +  );
 +
 +  $form['user']['ding_wayf_user_create_before_info']['content'] = array(
 +    '#type' => 'textarea',
 +    '#title' => t('Content'),
 +    '#default_value' => isset($defaults['content']) ? $defaults['content'] : '',
 +  );
 +
 +  $form['user']['ding_wayf_user_create_before_info']['link'] = array(
 +    '#type' => 'textfield',
 +    '#title' => t('Link title'),
 +    '#default_value' => isset($defaults['link']) ? $defaults['link'] : '',
 +  );
 +
 +  // Accepting text.
 +  $form['user']['ding_wayf_user_create_accept_text'] = array(
 +    '#type' => 'textarea',
 +    '#title' => t('Accept text'),
 +    '#default_value' => variable_get('ding_wayf_user_create_accept_text', ''),
 +    '#description' => t('The text that is display next to the user create accept check box.'),
 +  );
 +
 +  // Define after create info page.
 +  $form['user']['ding_wayf_user_create_after_info'] = array(
 +    '#type' => 'fieldset',
 +    '#tree' => TRUE,
 +    '#collapsible' => TRUE,
 +    '#collapsed' => TRUE,
 +    '#title' => t('Information page (after)'),
 +    '#description' => t('Information to display after the user have been create at the provider. This page can be used to display information about when the newly created user account is ready etc.'),
 +  );
 +
 +  $defaults = variable_get('ding_wayf_user_create_after_info', array());
 +  $form['user']['ding_wayf_user_create_after_info']['title'] = array(
 +    '#type' => 'textfield',
 +    '#title' => t('Title'),
 +    '#default_value' => isset($defaults['title']) ? $defaults['title'] : '',
 +  );
 +
 +  $form['user']['ding_wayf_user_create_after_info']['content'] = array(
 +    '#type' => 'textarea',
 +    '#title' => t('Content'),
 +    '#default_value' => isset($defaults['content']) ? $defaults['content'] : '',
 +  );
 +}
 +
 +/**
 + * Implements hook_theme().
 + */
 +function ding_wayf_dk_create_user_theme($existing, $type, $theme, $path) {
 +  return array(
 +    'ding_wayf_dk_info_page' => array(
 +      'template' => 'ding-wayf-dk-info-page',
 +      'variables' => array(
 +        'title' => array(),
 +        'content' => array(),
 +        'link' => NULL,
 +      ),
 +      'path' => $path . '/templates',
 +    ),
 +    'ding_wayf_dk_create_user_link' => array(
 +      'template' => 'ding-wayf-dk-create-user-link',
 +      'variables' => array(
 +        'link' => NULL,
 +      ),
 +      'path' => $path . '/templates',
 +    ),
 +  );
 +}
 +
 +/**
 + * Menu callback for information page before user creation.
 + */
 +function ding_wayf_dk_create_user_info_page() {
 +  // Get texts from the configuration.
 +  $values = variable_get('ding_wayf_user_create_before_info', array());
 +
 +  return array(
 +    '#theme' => 'ding_wayf_dk_info_page',
 +    '#title' => empty($values['title']) ? t('Not configured.') : $values['title'],
 +    '#content' => empty($values['content']) ? t('Please configure WAYF configuration to get content on this page under <a href="/admin/config/people/wayf">admin/config/people/wayf</a>.') : $values['content'],
@kasperg   Owner
kasperg added a note 10 days ago
Please provide a better default value. Having "Not configured" potentially exposed to end users is never a good idea.
 @kasperg   Owner
kasperg added a note 7 days ago
No admin instructions in end user texts. Please provide som sane defaults.
 @cableman   Owner
cableman added a note 4 days ago
I see, but it's hard to give any good defaults here... for aakb.dk it's a legal text and links to custom pages with the library rules etc.
 @cableman   Owner
cableman added a note 4 days ago
I see, but it's hard to give any good defaults here... for aakb.dk it's a legal text and links to custom pages with the library rules etc.
 Add a line note
 +    '#link' => array(
 +      '#theme' => 'link',
 +      '#text' => empty($values['link']) ? t('Register as new user') : $values['link'],
 +      '#path' => 'wayf/consume',
 +      '#options' => array(
 +        'attributes' => array(
 +          'class' => array(),
 +        ),
 +        'query' => array(
 +          'destination' => 'wayf/create-user/form',
 +        ),
 +        'html' => TRUE,
 +      ),
 +    ),
 +  );
 +}
 +
 +/**
 + * Menu callback for information page after user creation.
 + */
 +function ding_wayf_dk_create_user_notice() {
 +  // Get texts from the configuration.
 +  $values = variable_get('ding_wayf_user_create_after_info', array());
@kasperg   Owner
kasperg added a note 7 days ago
Naming seems inconsistent here. Some places we use notice - some places create after. Could this be made more consistent?
 Add a line note
 +
 +  return array(
 +    '#theme' => 'ding_wayf_dk_info_page',
 +    '#title' => empty($values['title']) ? t('Not configured.') : $values['title'],
 +    '#content' => empty($values['content']) ? t('Please configure WAYF configuration to get content on this page under <a href="/admin/config/people/wayf">admin/config/people/wayf</a>.') : $values['content'],
 +  );
 +}
 +
 +/**
 + * Menu callback to create the user creation form.
 + */
 +function ding_wayf_dk_create_user_form() {
 +  // Remove all message from the normal login, that have not succeed.
 +  drupal_get_messages(NULL, TRUE);
 +
 +  // Check if user have been logged in.
 +  if (user_is_logged_in()) {
 +    // User is logged in and have an account.
 +    drupal_set_message(t('You already have an user account at the library. So we have just logged you in.'));
 +
 +    // Ensure session information is removed.
 +    unset($_SESSION['ding_wayf_dk_create_user']);
 +    unset($_SESSION['ding_wayf_dk_login_destination']);
 +
 +    // Send user to front page.
 +    drupal_goto('<front>');
 +  }
 +
 +  // This is a little hack to get around the panels design requirements of the
 +  // ddbasic theme.
 +  $form = array(
 +    '#prefix' => '<div class="default-layout"><div class="layout-wrapper">',
 +    '#suffix' => '<br /></div></div>',
 +  );
@kasperg   Owner
kasperg added a note 7 days ago
I image this could be a problem that might occour multiple times. Could we extract it into a theme function and use #theme_wrappers?
 Add a line note
 +
 +  $form['accept'] = array(
 +    '#type' => 'checkbox',
 +    '#title' => t('I accept the user agreements'),
 +    '#description' => variable_get('ding_wayf_user_create_accept_text', ''),
 +    '#required' => TRUE,
 +  );
 +
 +  $form['name'] = array(
 +    '#type' => 'textfield',
 +    '#title' => t('Name'),
 +    '#description' => t('You full name.'),
@kasperg   Owner
kasperg added a note 7 days ago
Typo: Your.
 Add a line note
 +    '#maxlength' => 100,
@kasperg   Owner
kasperg added a note 7 days ago
Why this maxlength? If it is caused by other Drupal limitations then please add a comment stating why.